app/controllers/concerns/accounts/authorization.rb

```diff
@@ -0,0 +1,185 @@
+        raise <<-MESSAGE
+          Authorization check required for #{params[:action]} in #{self.class.name}.
+
+          Use any method of 'authorize', 'authorize_global', 'authorize_in_optional_project'
+          or 'require_admin' to ensure authorization. If authorization is checked by any other means,
+          affirm the same by calling 'authorization_checked!' in the controller. If the authorization does
+          not need to be checked for this action, affirm the same by calling 'no_authorization_required!'
+        MESSAGE
+      else
+        render_403
+      end
+    end
+  end
+
+  # Authorize the user for the requested controller action.
+  # To be used in before_action hooks
+  def authorize(ctrl = params[:controller], action = params[:action])
+    do_authorize({ controller: ctrl, action: }, global: false)
+  end
+
+  # Authorize the user for the requested controller action outside a project
+  # To be used in before_action hooks
+  def authorize_global
+    action = { controller: params[:controller], action: params[:action] }
+    do_authorize(action, global: true)
+  end
+
+  # Find a project based on params[:project_id]
+  def authorize_in_optional_project
+    @project = Project.find(params[:project_id]) if params[:project_id].present?
+
+    do_authorize({ controller: params[:controller], action: params[:action] }, global: params[:project_id].blank?)
+  rescue ActiveRecord::RecordNotFound
+    render_404
+  end
+
+  # Deny access if user is not allowed to do the specified action.
+  #
+  # Action can be:
+  # * a parameter-like Hash (eg. { controller: '/projects', action: 'edit' })
+  # * a permission Symbol (eg. :edit_project)
+  def do_authorize(action, global: false) # rubocop:disable Metrics/PerceivedComplexity
+    is_authorized = if global
+                      User.current.allowed_based_on_permission_context?(action)
+                    else
+                      User.current.allowed_based_on_permission_context?(action,
+                                                                        project: @project || @projects,
+                                                                        entity: @work_package || @work_packages)
+                    end
+
+    unless is_authorized
+      if @project&.archived?
+        render_403 message: :notice_not_authorized_archived_project
+      else
+        deny_access
+      end
+    end
+    is_authorized
+  end
+
+  def require_admin
+    return unless require_login
+
+    render_403 unless current_user.admin?
+  end
+
+  def authorization_is_ensured?(action)
+    return false if authorization_ensured.nil?
+
+    (authorization_ensured[:generally_allowed] == true || authorization_ensured[:only].include?(action.to_sym)) &&
+      authorization_ensured[:except].exclude?(action.to_sym)
+  end
+
+  class_methods do
+    # Overriding before_action of rails to check if any authorization method is by now defined.
+    def before_action(*names, &)
```

Comment:
Done in [7ba1c9a](https://github.com/opf/openproject/pull/15733/commits/7ba1c9a522427179f6a68d2c6c2f32f7d47d678a)

I didn't go with the after_action approach since I really want to be sure that nothing happens if authorization is not addressed.